setup/Core/SetupOperation.cs

```diff
@@ -151,44 +156,33 @@ private static bool DeleteEmptyDirsRecursion(string dir) {
 			return true;
 		}
 
-		protected readonly ITaskInterface taskInterface;
-		protected int progress;
-
-		protected SetupOperation(ITaskInterface taskInterface) {
-			this.taskInterface = taskInterface;
-		}
-
 		/// <summary>
-		/// Run the task, any exceptions thrown will be written to a log file and update the status label with the exception message
+		///     Run the task, any exceptions thrown will be written to a log file and update the status label with the exception
+		///     message
 		/// </summary>
-		public abstract void Run();
+		public abstract Task Run(IProgress progress, CancellationToken cancellationToken = default);
 
 		/// <summary>
-		/// Display a configuration dialog. Return false if the operation should be cancelled.
+		///     Display a configuration dialog. Return false if the operation should be cancelled.
 		/// </summary>
-		public virtual bool ConfigurationDialog() => true;
+		/// <param name="cancellationToken"></param>
+		public virtual ValueTask<bool> ConfigurationPrompt(CancellationToken cancellationToken = default) => ValueTask.FromResult(true);
 
 		/// <summary>
-		/// Display a startup warning dialog
+		///     Display a startup warning dialog
 		/// </summary>
 		/// <returns>true if the task should continue</returns>
 		public virtual bool StartupWarning() => true;
 
 		/// <summary>
-		/// Will prevent successive tasks from executing and cause FinishedDialog to be called
+		///     Will prevent successive tasks from executing and cause FinishedDialog to be called
 		/// </summary>
 		/// <returns></returns>
 		public virtual bool Failed() => false;
 
 		/// <summary>
-		/// Will cause FinishedDialog to be called if warnings are not supressed
-		/// </summary>
-		/// <returns></returns>
-		public virtual bool Warnings() => false;
-
-		/// <summary>
-		/// Called to display a finished dialog if Failures() || warnings are not supressed and Warnings()
+		///     Called to display a finished dialog if Failures() || warnings are not supressed and Warnings()
 		/// </summary>
-		public virtual void FinishedDialog() {}
+		public virtual void FinishedPrompt() { }
```

Comment:
I wonder if we can get rid of `FinishedPrompt` and `Failed` entirely, and just use exceptions and `Inform` in the `Run` methods. 

I can't remember why I had `FinishedPrompt` split out originally. Probably overzealous implementation of `CompositeTask`. I was much greener 9 years ago.

Reply:
This is actually required because of what I mentioned here: https://github.com/tModLoader/tModLoader/pull/4353#discussion_r1734974285